x/fundraising/types/utils.go

```diff
@@ -34,8 +34,98 @@ func DeriveAddress(addressType AddressType, moduleName, name string) sdk.AccAddr
 
 // SortByBidPrice sorts bid array by bid price in descending order.
 func SortByBidPrice(bids []Bid) []Bid {
-	sort.SliceStable(bids, func(i, j int) bool {
-		return bids[i].Price.GT(bids[j].Price)
+	sort.Slice(bids, func(i, j int) bool {
+		if bids[i].Price.GT(bids[j].Price) {
+			return true
+		}
+		return bids[i].Id < bids[j].Id
 	})
 	return bids
 }
+
+func BidsByPrice(bids []Bid) (prices []sdk.Dec, bidsByPrice map[string][]Bid) {
+	bids = SortByBidPrice(bids)
+
+	bidsByPrice = map[string][]Bid{} // price => []Bid
+
+	for _, bid := range bids {
+		priceStr := bid.Price.String()
+		bidsByPrice[priceStr] = append(bidsByPrice[priceStr], bid)
+	}
+
+	// Sort prices in descending order.
+	prices = make([]sdk.Dec, len(bidsByPrice))
+	i := 0 // TODO: is it too much optimization? we can use append(...)
+	for priceStr := range bidsByPrice {
+		prices[i] = sdk.MustNewDecFromStr(priceStr)
+		i++
+	}
+	sort.Slice(prices, func(i, j int) bool {
+		return prices[i].GT(prices[j])
+	})
+	return
+}
+
+type MatchResult struct {
+	MatchPrice          sdk.Dec
+	MatchedAmount       sdk.Int
+	MatchedBids         []Bid
+	MatchResultByBidder map[string]*BidderMatchResult
+}
+
+type BidderMatchResult struct {
+	PayingAmount  sdk.Int
+	MatchedAmount sdk.Int
+}
+
+func Match(auction AuctionI, matchPrice sdk.Dec, prices []sdk.Dec, bidsByPrice map[string][]Bid) (res *MatchResult, matched bool) {
+	biddableAmtByBidder := auction.GetAllowedBiddersMap()
+	res = &MatchResult{
+		MatchPrice:          matchPrice,
+		MatchedAmount:       sdk.ZeroInt(),
+		MatchResultByBidder: map[string]*BidderMatchResult{},
+	}
+
+	for _, price := range prices {
+		if price.LT(matchPrice) {
+			break
+		}
+
+		for _, bid := range bidsByPrice[price.String()] {
+			var bidAmt sdk.Int
+			switch bid.Type {
+			case BidTypeBatchWorth:
+				bidAmt = bid.Coin.Amount.ToDec().QuoTruncate(matchPrice).TruncateInt()
```

Comment:
We can use `bid.ConvertToPayingAmount(denom string)` method here.

Reply:
I used the formula directly because the bidding amount should be calculated based on `matchPrice`, not on its own `Bid.Price`.

Reply:
Oh, yes. You're right. I got it.